### ISSUE TITLE
Analysis identifies how to reach null

### ISSUE THREAD

**Issue by ErliSoares** [dart-lang/language#1500]
Error analysis identifies the expression as an error, but under no conditions would the expression be null.

```dart
  static bool isEqual(double? value, double? valueCompare) {
    if (value == null && valueCompare != null) {
      return false;
    }
    if (valueCompare == null && value != null) {
      return false;
    }
    if (valueCompare == null && value == null) {
      return true;
    }
    return double.parse(value.toStringAsFixed(2)) == double.parse(valueCompare.toStringAsFixed(2));
  }
```

Message: ```error: The method 'toStringAsFixed' can't be unconditionally invoked because the receiver can be 'null'. (unchecked_use_of_nullable_value at [...] ..:...)```

**Comment by leafpetersen** [comment]
@stereotype441 thoughts? At first glance, this would seem like a heavy lift to me, but maybe it falls into the sort of thing your correlated variables analysis could track?

**Comment by stereotype441** [comment]
> @stereotype441 thoughts? At first glance, this would seem like a heavy lift to me, but maybe it falls into the sort of thing your correlated variables analysis could track?

Yeah, unfortunately supporting this sort of thing would require a pretty major re-think of flow analysis.  And I worry that if we weren't super careful about it we would risk accidentally creating an algorithm that's O(2^n) in the number of local variables in the function.
